specs/interop/liquidity_migration.md

```diff
@@ -0,0 +1,358 @@
+
+```solidity
+convertToSuper(address _legacyAddr, address _superAddr, uint256 _amount)
+```
+
+The function will
+
+1. Check that `_legacyAddr` and `_superAddr` are valid and paired.
+2. Check both tokens have the same amount of decimals.
+3. Burn `_amount` of `_legacyAddr` from `msg.sender`.
+4. Mint `_amount` of `_superAddr` to `msg.sender`.
+
+An example implementation that depends on the `checkPair()`
+access control function looks like this:
+
+```solidity
+function convertToSuper(address _legacyAddr, address _superAddr, uint256 _amount) public {
+  require(checkPair(_legacyAddr, _superAddr), "Invalid address pair");
+  require(IERC20(_legacyAddr).decimals() == IERC20(_superAddr).decimals(), "Decimals do not match")
+
+  IERC20(_legacyAddr).burn(msg.sender, _amount);
+  IERC20(_superAddr).mint(msg.sender, _amount);
+
+  emit ConvertedToSuper(_legacyAddr, _superAddr, msg.sender, _amount);
+}
+```
+
+#### `convertFromSuper`
+
+Converts `_amount` of `SuperchainERC20` tokens in address `_superAddr`
+to the same `_amount` of legacy tokens with address `_legacyAddr`.
+
+```solidity
+convertToSuper(address _legacyAddr, address _superAddr, uint256 _amount)
+```
+
+The function will
+
+1. Check that `_legacyAddr` and `_superAddr` are valid and paired.
+2. Check both tokens have the same amount of decimals.
+3. Burn `_amount` of `_superAddr` from `msg.sender`.
+4. Mint `_amount` of `_legacyAddr` to `msg.sender`.
+
+An example implementation that depends on the `checkPair()`
+access control function looks like this:
+
+```solidity
+function convertFromSuper(address _legacyAddr, address _superAddr, uint256 _amount) public {
+  require(checkPair(_legacyAddr, _superAddr), "Invalid address pair");
+  require(IERC20(_legacyAddr).decimals() == IERC20(_superAddr).decimals(), "Decimals do not match")
+
+  IERC20(_superAddr).burn(msg.sender, _amount);
+  IERC20(_legacyAddr).mint(msg.sender, _amount);
+
+  emit ConvertedFromSuper(_legacyAddr, _superAddr, msg.sender, _amount);
+}
+```
+
+### Events
+
+#### `ConvertedToSuper`
+
+It MUST trigger when anyone converts legacy tokens
+to `SuperchainERC20` with `convertToSuper`.
+
+```solidity
+event ConvertedToSuper(address indexed legacyAddr, address indexed superAddr, address indexed caller, uint256 amount, uint256 adjustedAmount);
+```
+
+#### `ConvertedFromSuper`
+
+It MUST trigger when anyone converts `SuperchainERC20`
+into legacy tokens with `convertFromSuper`.
+
+```solidity
+event ConvertedFromSuper(address indexed legacyAddr, address indexed superAddr, address indexed caller, uint256 amount);
+```
+
+### Internal Functions
+
+#### `checkPair`
+
+Checks the validity of the `_legacyAddr` and `_superAddr` pair.
+To do so, it will verify:
+
+1. Valid `OptimismMintableERC20` address:
+
+   1. For representations minted before
+      the `OptimismMintableERC20Factory` upgrade:
+      Check an allowed token list that connects L2 representations with the
```

Comment:
It would be relatively straightforward to do this trustlessly but would require that we have an external function that can be called to populate the mapping. I would prefer that over a trusted solution.

Reply:
@smartcontracts the idea was to do it in a trusted manner, here's a relevant comment from @tynes: https://github.com/ethereum-optimism/design-docs/pull/46#discussion_r1678444097, proto had suggested a slightly different approach using an onion-hash, which is the R&D discord

Reply:
We can assume that the mapping will be populated correctly as part of this spec, we can work on a design doc for how to populate the mapping separately